#  **Sistema de Recomendación**

Buscamos los business_id que representan una buena estrategia de inversión en el dataset inicial para extraer:

1. Services.
2. Attributes.
3. City.
4. State.

Y juntarlos en una sola columna.

Filtrado por:
1. Categories.

Vamos a usar el dataframe combinado de Google Maps y Yelp.

In [171]:
import pandas as pd

# Set the option to display all columns
pd.set_option('display.max_columns', None)

df = pd.read_csv('df_final0611.csv')
df

business_id                business_name  \
0                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
1                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
2                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
3                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
4                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
...                                       ...                          ...   
915353  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   
915354  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   
915355  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   
915356  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   
915357  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   

                                                  address       city  \
0                                            no direction   Meridian   
1                                            no direction   Meridian   
2                                            no direction   Meridian   
3                                            no direction   Meridian   
4                                            no direction   Meridian   
...                                                   ...        ...   
915353  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   
915354  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   
915355  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   
915356  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   
915357  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   

             state   latitude   longitude  stars_business  review_count  \
0          Florida  43.643494 -116.436000             5.0             5   
1          Florida  43.643494 -116.436000             5.0             5   
2          Florida  43.643494 -116.436000             5.0             5   
3          Florida  43.643494 -116.436000             5.0             5   
4          Florida  43.643494 -116.436000             5.0             5   
...            ...        ...         ...             ...           ...   
915353  New_Mexico  35.177837 -103.725132             4.1            13   
915354  New_Mexico  35.177837 -103.725132             4.1            13   
915355  New_Mexico  35.177837 -103.725132             4.1            13   
915356  New_Mexico  35.177837 -103.725132             4.1            13   
915357  New_Mexico  35.177837 -103.725132             4.1            13   

                               categories  \
0       Local Services Appliances  Repair   
1       Local Services Appliances  Repair   
2       Local Services Appliances  Repair   
3       Local Services Appliances  Repair   
4       Local Services Appliances  Repair   
...                                   ...   
915353              Glass cutting service   
915354              Glass cutting service   
915355              Glass cutting service   
915356              Glass cutting service   
915357              Glass cutting service   

                                               attributes  \
0                           Business Accepts Credit Cards   
1                           Business Accepts Credit Cards   
2                           Business Accepts Credit Cards   
3                           Business Accepts Credit Cards   
4                           Business Accepts Credit Cards   
...                                                   ...   
915353  'In-store shopping' 'Wheelchair accessible ent...   
915354  'In-store shopping' 'Wheelchair accessible ent...   
915355  'In-store shopping' 'Wheelchair accessible ent...   
915356  'In-store shopping' 'Wheelchair accessible ent...   
915357  'In-store shopping' 'Wheelchair accessible ent...   

                       user_id  stars_by_user  \
0       

Nos aseguramos de que en la columna 'state' se reemplacen los '_' por ' '.

In [172]:
df['state'] = df['state'].str.replace('_', ' ')
df

business_id                business_name  \
0                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
1                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
2                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
3                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
4                      n7AQvGvNHlmun3kqXeBKVQ      Roy's Appliance Service   
...                                       ...                          ...   
915353  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   
915354  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   
915355  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   
915356  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   
915357  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari Lumber True Value   

                                                  address       city  \
0                                            no direction   Meridian   
1                                            no direction   Meridian   
2                                            no direction   Meridian   
3                                            no direction   Meridian   
4                                            no direction   Meridian   
...                                                   ...        ...   
915353  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   
915354  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   
915355  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   
915356  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   
915357  Tucumcari Lumber True Value, 221 S 1st St, Tuc...  Tucumcari   

             state   latitude   longitude  stars_business  review_count  \
0          Florida  43.643494 -116.436000             5.0             5   
1          Florida  43.643494 -116.436000             5.0             5   
2          Florida  43.643494 -116.436000             5.0             5   
3          Florida  43.643494 -116.436000             5.0             5   
4          Florida  43.643494 -116.436000             5.0             5   
...            ...        ...         ...             ...           ...   
915353  New Mexico  35.177837 -103.725132             4.1            13   
915354  New Mexico  35.177837 -103.725132             4.1            13   
915355  New Mexico  35.177837 -103.725132             4.1            13   
915356  New Mexico  35.177837 -103.725132             4.1            13   
915357  New Mexico  35.177837 -103.725132             4.1            13   

                               categories  \
0       Local Services Appliances  Repair   
1       Local Services Appliances  Repair   
2       Local Services Appliances  Repair   
3       Local Services Appliances  Repair   
4       Local Services Appliances  Repair   
...                                   ...   
915353              Glass cutting service   
915354              Glass cutting service   
915355              Glass cutting service   
915356              Glass cutting service   
915357              Glass cutting service   

                                               attributes  \
0                           Business Accepts Credit Cards   
1                           Business Accepts Credit Cards   
2                           Business Accepts Credit Cards   
3                           Business Accepts Credit Cards   
4                           Business Accepts Credit Cards   
...                                                   ...   
915353  'In-store shopping' 'Wheelchair accessible ent...   
915354  'In-store shopping' 'Wheelchair accessible ent...   
915355  'In-store shopping' 'Wheelchair accessible ent...   
915356  'In-store shopping' 'Wheelchair accessible ent...   
915357  'In-store shopping' 'Wheelchair accessible ent...   

                       user_id  stars_by_user  \
0       

Revisamos las columnas:

In [173]:
columns = df.columns
print(columns)

Index(['business_id', 'business_name', 'address', 'city', 'state', 'latitude',
       'longitude', 'stars_business', 'review_count', 'categories',
       'attributes', 'user_id', 'stars_by_user', 'review_text', 'review_date'],
      dtype='object')


Eliminamos las columnas que no nos van a ser útiles en el sistema de recomendación:

In [174]:
df1 = df[['business_id', 'city', 'state', 'stars_business', 'review_count', 'categories','attributes']]
df1

business_id       city       state  \
0                      n7AQvGvNHlmun3kqXeBKVQ   Meridian     Florida   
1                      n7AQvGvNHlmun3kqXeBKVQ   Meridian     Florida   
2                      n7AQvGvNHlmun3kqXeBKVQ   Meridian     Florida   
3                      n7AQvGvNHlmun3kqXeBKVQ   Meridian     Florida   
4                      n7AQvGvNHlmun3kqXeBKVQ   Meridian     Florida   
...                                       ...        ...         ...   
915353  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari  New Mexico   
915354  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari  New Mexico   
915355  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari  New Mexico   
915356  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari  New Mexico   
915357  0x871b577e95cb04b5:0x4829567f16096b2e  Tucumcari  New Mexico   

        stars_business  review_count                         categories  \
0                  5.0             5  Local Services Appliances  Repair   
1                  5.0             5  Local Services Appliances  Repair   
2                  5.0             5  Local Services Appliances  Repair   
3                  5.0             5  Local Services Appliances  Repair   
4                  5.0             5  Local Services Appliances  Repair   
...                ...           ...                                ...   
915353             4.1            13              Glass cutting service   
915354             4.1            13              Glass cutting service   
915355             4.1            13              Glass cutting service   
915356             4.1            13              Glass cutting service   
915357             4.1            13              Glass cutting service   

                                               attributes  
0                           Business Accepts Credit Cards  
1                           Business Accepts Credit Cards  
2                           Business Accepts Credit Cards  
3                           Business Accepts Credit Cards  
4                           Business Accepts Credit Cards  
...                                                   ...  
915353  'In-store shopping' 'Wheelchair accessible ent...  
915354  'In-store shopping' 'Wheelchair accessible ent...  
915355  'In-store shopping' 'Wheelchair accessible ent...  
915356  'In-store shopping' 'Wheelchair accessible ent...  
915357  'In-store shopping' 'Wheelchair accessible ent...  

[915358 rows x 7 columns]

Ahora sacamos el dataframe usando los unique business_id:

In [218]:
# Eliminar duplicados basados en la columna 'business_id' y mantener la primera fila de cada 'business_id'
dfg = df1.drop_duplicates(subset='business_id', keep='first').reset_index(drop=True)
dfg

business_id              city         state  \
0                     n7AQvGvNHlmun3kqXeBKVQ          Meridian       Florida   
1                     ZM46RDLXaFNo_z6t-j_L4w          Bensalem    California   
2                     2n9HHBxG7yAyAUwXXa49aw       Southampton      Delaware   
3                     yv6kSKxDdZTsvHL25K-U2g         Nashville   Pensilvania   
4                     DTUfAthScapCGnOnnLUhPA          Glenside     Tennessee   
...                                      ...               ...           ...   
41760  0x89b2d96db3d2f989:0xaeab3070f91b2c4a   Madison Heights      Virginia   
41761  0x89b2d04902f17fc9:0x9bc7db53b3b3f8f0         Lynchburg      Virginia   
41762  0x8804757a4d9d6d4d:0xd1c6cd2fb9aed070       Fond du Lac     Wisconsin   
41763   0x89e446925bd5e4a7:0xc1d00e81eb1dc17  North Smithfield  Rhode Island   
41764  0x871b577e95cb04b5:0x4829567f16096b2e         Tucumcari    New Mexico   

       stars_business  review_count  \
0                 5.0             5   
1                 3.5             5   
2                 4.0             5   
3                 4.5            45   
4                 3.5             9   
...               ...           ...   
41760             4.7           115   
41761             4.1            68   
41762             4.4            18   
41763             4.4            38   
41764             4.1            13   

                                              categories  \
0                      Local Services Appliances  Repair   
1      MasonryConcrete Gardeners Lawn Services Tree S...   
2            Flooring Handyman Contractors Home Services   
3      Painters Home Services Decks  Railing Contract...   
4      Art Classes Shopping Education Arts  Crafts Ap...   
...                                                  ...   
41760  HVAC contractor Air conditioning contractor El...   
41761               Chimney sweep Hot tub repair service   
41762                                            Plumber   
41763                          Lawn mower repair service   
41764                              Glass cutting service   

                                              attributes  
0                          Business Accepts Credit Cards  
1                          Business Accepts Credit Cards  
2                          Business Accepts Credit Cards  
3       Business Accepts Credit Cards  By Appointment...  
4            Business Accepts Credit Cards  Bike Parking  
...                                                  ...  
41760  'Onsite services' 'Online estimates' 'Appointm...  
41761  'Curbside pickup' 'In-store pickup' 'In-store ...  
41762                                                NaN  
41763  'Curbside pickup' 'In-store pickup' 'In-store ...  
41764  'In-store shopping' 'Wheelchair accessible ent...  

[41765 rows x 7 columns]

Cambiamos el nombre de la columna 'categories' a 'services'

In [219]:
# Cambiar el nombre de la columna 'categories' a 'services' usando la función 'rename'
dfg = dfg.rename(columns={'categories': 'services'})
dfg

business_id              city         state  \
0                     n7AQvGvNHlmun3kqXeBKVQ          Meridian       Florida   
1                     ZM46RDLXaFNo_z6t-j_L4w          Bensalem    California   
2                     2n9HHBxG7yAyAUwXXa49aw       Southampton      Delaware   
3                     yv6kSKxDdZTsvHL25K-U2g         Nashville   Pensilvania   
4                     DTUfAthScapCGnOnnLUhPA          Glenside     Tennessee   
...                                      ...               ...           ...   
41760  0x89b2d96db3d2f989:0xaeab3070f91b2c4a   Madison Heights      Virginia   
41761  0x89b2d04902f17fc9:0x9bc7db53b3b3f8f0         Lynchburg      Virginia   
41762  0x8804757a4d9d6d4d:0xd1c6cd2fb9aed070       Fond du Lac     Wisconsin   
41763   0x89e446925bd5e4a7:0xc1d00e81eb1dc17  North Smithfield  Rhode Island   
41764  0x871b577e95cb04b5:0x4829567f16096b2e         Tucumcari    New Mexico   

       stars_business  review_count  \
0                 5.0             5   
1                 3.5             5   
2                 4.0             5   
3                 4.5            45   
4                 3.5             9   
...               ...           ...   
41760             4.7           115   
41761             4.1            68   
41762             4.4            18   
41763             4.4            38   
41764             4.1            13   

                                                services  \
0                      Local Services Appliances  Repair   
1      MasonryConcrete Gardeners Lawn Services Tree S...   
2            Flooring Handyman Contractors Home Services   
3      Painters Home Services Decks  Railing Contract...   
4      Art Classes Shopping Education Arts  Crafts Ap...   
...                                                  ...   
41760  HVAC contractor Air conditioning contractor El...   
41761               Chimney sweep Hot tub repair service   
41762                                            Plumber   
41763                          Lawn mower repair service   
41764                              Glass cutting service   

                                              attributes  
0                          Business Accepts Credit Cards  
1                          Business Accepts Credit Cards  
2                          Business Accepts Credit Cards  
3       Business Accepts Credit Cards  By Appointment...  
4            Business Accepts Credit Cards  Bike Parking  
...                                                  ...  
41760  'Onsite services' 'Online estimates' 'Appointm...  
41761  'Curbside pickup' 'In-store pickup' 'In-store ...  
41762                                                NaN  
41763  'Curbside pickup' 'In-store pickup' 'In-store ...  
41764  'In-store shopping' 'Wheelchair accessible ent...  

[41765 rows x 7 columns]

Revisamos cómo están organizados los valores de la columna 'services':

In [220]:
unique_services = dfg['services'].unique()
print(unique_services)

['Local Services Appliances  Repair'
 'MasonryConcrete Gardeners Lawn Services Tree Services Landscape Architects Home Services Contractors Landscaping'
 'Flooring Handyman Contractors Home Services' ...
 'Landscaper Landscape designer Lawn care service Lawn sprinkler system contractor Snow removal service Tree service'
 'Bathroom remodeler Carpenter Dry wall contractor Fence contractor Kitchen remodeler Pressure washing service Railing contractor Wood floor installation service Wood floor refinishing service'
 'HVAC contractor Air conditioning contractor Electrician Furnace repair service Heating contractor']


In [221]:
unique_attributes = dfg['attributes'].unique()
print(unique_attributes)

[' Business Accepts Credit Cards'
 ' Business Accepts Credit Cards  By Appointment Only'
 ' Business Accepts Credit Cards  Bike Parking' ...
 "'Onsite services' 'Online estimates' 'Appointment required' 'Staff wear masks' 'Repair services'"
 "'Curbside pickup' 'In-store pickup' 'In-store shopping' 'Mask required' 'Staff wear masks' 'Staff required to disinfect surfaces between visits' 'Safety dividers at checkout' 'Repair services' 'Quick visit'"
 "'In-store shopping' 'Wheelchair accessible entrance' 'Assembly service' 'Repair services' 'Service guarantee' 'Quick visit' 'NFC mobile payments'"]


Vamos a cambiar los valores NA de estas columnas por ''.

In [222]:
# Filtrar las filas que contienen NaN en la columna 'services'
na_values = dfg[dfg['attributes'].isna()]
na_values

business_id           city           state  \
8169   0x87bf6f4475185551:0x4fca89e37cf34728       Lawrence          Kansas   
8170   0x876cad7309d758e7:0x6b03aacbbbf78528       Monument        Colorado   
8172   0x809a61887b1c50b1:0xef8d3aacbe2d0df8       Plymouth      California   
8175   0x876c6fbd63d70b0b:0xd6cf66e5f18bb074  Commerce City        Colorado   
8176   0x876c8dcc7187348b:0xf561fa5a6039a910         Aurora        Colorado   
...                                      ...            ...             ...   
41751  0x54904291c64dff39:0x2388530f1373e968        Seattle      Washington   
41752  0x885301a983710559:0xaa63ca1ae9eac3c1   Kernersville  North Carolina   
41753  0x872272765a810293:0x598d0c7e6df6e352    Albuquerque      New Mexico   
41757  0x8853072e0a69152f:0x92f929bd94a72834         Colfax  North Carolina   
41762  0x8804757a4d9d6d4d:0xd1c6cd2fb9aed070    Fond du Lac       Wisconsin   

       stars_business  review_count  \
8169              4.9            87   
8170              4.3             8   
8172              4.6            18   
8175              4.3             6   
8176              3.9             8   
...               ...           ...   
41751             3.8            78   
41752             4.9           473   
41753             4.8           106   
41757             4.8           168   
41762             4.4            18   

                                                services attributes  
8169                                        Tree service        NaN  
8170                                        Home builder        NaN  
8172                                        Home builder        NaN  
8175                                        Home builder        NaN  
8176                                        Home builder        NaN  
...                                                  ...        ...  
41751  Appliance repair service Refrigerator repair s...        NaN  
41752  Air conditioning contractor Air conditioning r...        NaN  
41753  Appliance repair service Refrigerator repair s...        NaN  
41757  Water damage restoration service Fire damage r...        NaN  
41762                                            Plumber        NaN  

[11802 rows x 7 columns]

In [223]:
dfg['attributes'] = dfg['attributes'].fillna('')

El sistema de recomendación lo vamos a hacer con la combinación de 'services' y 'attributes' para esto creamos un nuevo data set que tenga el business_id y la combinación de estas dos columnas.

In [224]:
columns = dfg.columns
print(columns)

Index(['business_id', 'city', 'state', 'stars_business', 'review_count',
       'services', 'attributes'],
      dtype='object')


Vemos que para los negocios de Yelp, los 'services' no aparecen delimitados por '', sino que cada service empieza por mayúsculas y las demás palabras empiezan en minúsculas. Vamos a crear un método que me ponga las '' a cada frase.

In [225]:
dfg.iloc[41757,5]

'Water damage restoration service Fire damage restoration service'

Vamos a revisar las filas donde aparece HVAC en la columna service y miramos las frases únicas.

In [226]:
unique_hvac_services = dfg[dfg['services'].str.contains('HVAC', case=False)]['services'].unique()
unique_hvac_services

array(['Home Services Heating  Air ConditioningHVAC',
       'Home Services Home Cleaning Office Cleaning Local Services Handyman Heating  Air ConditioningHVAC Carpet Cleaning Pressure Washers Professional Services',
       'Heating  Air ConditioningHVAC Home Services', ...,
       'HVAC contractor Air conditioning repair service Electrical installation service Electrician Heating contractor Plumber',
       'HVAC contractor Air conditioning contractor Air conditioning repair service Appliance repair service Furnace repair service Heating contractor Insulation contractor Solar energy contractor',
       'HVAC contractor Air conditioning contractor Electrician Furnace repair service Heating contractor'],
      dtype=object)

In [227]:
import re

# Define una función para separar las frases en comillas simples
def separate_phrases(text):
    phrases = re.findall(r"[A-Z][a-z\s']+", text)
    return " ".join(f"'{phrase}'" for phrase in phrases)

# Aplica la función a la columna 'services'
dfg['services'] = dfg['services'].apply(separate_phrases)

# Imprime el DataFrame resultante
dfg

business_id              city         state  \
0                     n7AQvGvNHlmun3kqXeBKVQ          Meridian       Florida   
1                     ZM46RDLXaFNo_z6t-j_L4w          Bensalem    California   
2                     2n9HHBxG7yAyAUwXXa49aw       Southampton      Delaware   
3                     yv6kSKxDdZTsvHL25K-U2g         Nashville   Pensilvania   
4                     DTUfAthScapCGnOnnLUhPA          Glenside     Tennessee   
...                                      ...               ...           ...   
41760  0x89b2d96db3d2f989:0xaeab3070f91b2c4a   Madison Heights      Virginia   
41761  0x89b2d04902f17fc9:0x9bc7db53b3b3f8f0         Lynchburg      Virginia   
41762  0x8804757a4d9d6d4d:0xd1c6cd2fb9aed070       Fond du Lac     Wisconsin   
41763   0x89e446925bd5e4a7:0xc1d00e81eb1dc17  North Smithfield  Rhode Island   
41764  0x871b577e95cb04b5:0x4829567f16096b2e         Tucumcari    New Mexico   

       stars_business  review_count  \
0                 5.0             5   
1                 3.5             5   
2                 4.0             5   
3                 4.5            45   
4                 3.5             9   
...               ...           ...   
41760             4.7           115   
41761             4.1            68   
41762             4.4            18   
41763             4.4            38   
41764             4.1            13   

                                                services  \
0           'Local ' 'Services ' 'Appliances  ' 'Repair'   
1      'Masonry' 'Concrete ' 'Gardeners ' 'Lawn ' 'Se...   
2      'Flooring ' 'Handyman ' 'Contractors ' 'Home '...   
3      'Painters ' 'Home ' 'Services ' 'Decks  ' 'Rai...   
4      'Art ' 'Classes ' 'Shopping ' 'Education ' 'Ar...   
...                                                  ...   
41760  'C contractor ' 'Air conditioning contractor '...   
41761          'Chimney sweep ' 'Hot tub repair service'   
41762                                          'Plumber'   
41763                        'Lawn mower repair service'   
41764                            'Glass cutting service'   

                                              attributes  
0                          Business Accepts Credit Cards  
1                          Business Accepts Credit Cards  
2                          Business Accepts Credit Cards  
3       Business Accepts Credit Cards  By Appointment...  
4            Business Accepts Credit Cards  Bike Parking  
...                                                  ...  
41760  'Onsite services' 'Online estimates' 'Appointm...  
41761  'Curbside pickup' 'In-store pickup' 'In-store ...  
41762                                                     
41763  'Curbside pickup' 'In-store pickup' 'In-store ...  
41764  'In-store shopping' 'Wheelchair accessible ent...  

[41765 rows x 7 columns]

Vemos que hay lugares en los que quedaron como 'C contractor' y debe ser 'HVAC contractor'. Vamos a hcer este cambio.

In [228]:
# Crear una nueva columna 'combSA' combinando 'services' y 'attributes'
dfg['combSA'] = dfg['services'].astype(str) + ' ' + dfg['attributes'].astype(str)
# Crear una nueva columna 'city_state' combinando 'city' y 'state'
dfg['city_state'] = dfg['city'].astype(str) + ', ' + dfg['state'].astype(str)

# Seleccionar las columnas 'business_id' y 'combSA' en un nuevo DataFrame
dfrsSA = dfg[['business_id', 'city_state', 'combSA']]

# Mostrar el nuevo DataFrame
dfrsSA

business_id                      city_state  \
0                     n7AQvGvNHlmun3kqXeBKVQ               Meridian, Florida   
1                     ZM46RDLXaFNo_z6t-j_L4w            Bensalem, California   
2                     2n9HHBxG7yAyAUwXXa49aw           Southampton, Delaware   
3                     yv6kSKxDdZTsvHL25K-U2g          Nashville, Pensilvania   
4                     DTUfAthScapCGnOnnLUhPA             Glenside, Tennessee   
...                                      ...                             ...   
41760  0x89b2d96db3d2f989:0xaeab3070f91b2c4a       Madison Heights, Virginia   
41761  0x89b2d04902f17fc9:0x9bc7db53b3b3f8f0             Lynchburg, Virginia   
41762  0x8804757a4d9d6d4d:0xd1c6cd2fb9aed070          Fond du Lac, Wisconsin   
41763   0x89e446925bd5e4a7:0xc1d00e81eb1dc17  North Smithfield, Rhode Island   
41764  0x871b577e95cb04b5:0x4829567f16096b2e           Tucumcari, New Mexico   

                                                  combSA  
0      'Local ' 'Services ' 'Appliances  ' 'Repair'  ...  
1      'Masonry' 'Concrete ' 'Gardeners ' 'Lawn ' 'Se...  
2      'Flooring ' 'Handyman ' 'Contractors ' 'Home '...  
3      'Painters ' 'Home ' 'Services ' 'Decks  ' 'Rai...  
4      'Art ' 'Classes ' 'Shopping ' 'Education ' 'Ar...  
...                                                  ...  
41760  'C contractor ' 'Air conditioning contractor '...  
41761  'Chimney sweep ' 'Hot tub repair service' 'Cur...  
41762                                         'Plumber'   
41763  'Lawn mower repair service' 'Curbside pickup' ...  
41764  'Glass cutting service' 'In-store shopping' 'W...  

[41765 rows x 3 columns]

Vamos a guardar este dataframe en un archivo .csv que vamos a usar en el archivo main.py de la API ya deployada

In [409]:
dfrsSA.to_csv('mejores_business.csv', index=False)

De estos 41765 negocios únicos, vamos a obtener los que tienen un mejor performance y usar solo estos para desarrollar el sistema de recomendación. Queremos asegurarnos de recomendar lo mejor a nuestro cliente Construction Valdes. Para eso vamos a traer un dataset con todos los negocios clasificados bajo unos criterios que explicaremos a continuación:

In [410]:
dfcrit = pd.read_csv('merged_stat_df.csv')
dfcrit

business_id        city  \
0      0x888d5f9bb1c9ef87:0x19aa747cabb74b93   Abbeville   
1      0x8889266653174d8d:0xb719a2e60a518428   Alabaster   
2      0x8888dfab85845773:0x3891db3a54888d54   Alabaster   
3      0x888925cc876b0023:0x56c38bc00d9545b2   Alabaster   
4      0x88892f9e8dead30b:0x8dccd8514867355c   Alabaster   
...                                      ...         ...   
21578  0x89b7b9af2e9109a3:0x87443f600f805ad0  Washington   
21579  0x89b7c868b628976d:0xf143ff5930a27d36  Washington   
21580  0x89b7b60791c1d4bd:0x306b5fb54c95180d  Washington   
21581  0x89b7b7bbcfd7a0eb:0x26033c1d91ba1091  Washington   
21582  0x89b7b8333a858d8b:0x60352910ef015906  Washington   

                      state                           categories  \
0                   Alabama      Appliance Repair & Installation   
1                   Alabama                    Cleaning Services   
2                   Alabama                  Interior Remodeling   
3                   Alabama                   HVAC & Air Quality   
4                   Alabama                   HVAC & Air Quality   
...                     ...                                  ...   
21578  District of Columbia      Appliance Repair & Installation   
21579  District of Columbia                    Cleaning Services   
21580  District of Columbia  Exterior Remodeling or Construction   
21581  District of Columbia                   HVAC & Air Quality   
21582  District of Columbia                    Handyman or Other   

       high_category_consumption  low_category_volume  high_city_state_income  \
0                          False                False                   False   
1                           True                False                    True   
2                          False                False                    True   
3                           True                 True                    True   
4                           True                 True                    True   
...                          ...                  ...                     ...   
21578                      False                False                   False   
21579                      False                False                   False   
21580                      False                False                   False   
21581                      False                False                   False   
21582                      False                False                   False   

       high_star_business  high_review_count  
0                   False              False  
1                    True               True  
2                   False              False  
3                   False              False  
4                   False               True  
...                   ...                ...  
21578                True               True  
21579               False              False  
21580                True               True  
21581               False               True  
21582               False              False  

[21583 rows x 9 columns]

Revisamos las categorias y las guardamos en una variable llamada 'unique_categories':

In [411]:
unique_categories = dfcrit['categories'].unique()
print(unique_categories)

['Appliance Repair & Installation' 'Cleaning Services'
 'Interior Remodeling' 'HVAC & Air Quality' 'Handyman or Other'
 'Landscape & Garden Services' 'Exterior Remodeling or Construction'
 'Roofing & Insulation' 'Plumbing Services' 'Masonry & Concrete Services'
 'Electrical Services' 'Home Inspection & Real Estate'
 'Art & Enterteinment']


Vamos a hacer un primer sistema de recomendación para recomendar servicios y atributos basados en en el contenido. Vamos a tomar como muestra los business_id que tienen altas calificaciones y altos reviews, es decir que son negocios con un buen desempeño. 

In [414]:
mask1 = (dfcrit['high_star_business'] == True) & (dfcrit['high_review_count'] == True) & (dfcrit['categories'] == 'Art & Enterteinment')
bid_rsSA = dfcrit[mask1]
bid_rsSA

Empty DataFrame
Columns: [business_id, city, state, categories, high_category_consumption, low_category_volume, high_city_state_income, high_star_business, high_review_count]
Index: []

In [233]:
dfrsSA_filtered = dfrsSA[dfrsSA['business_id'].isin(bid_rsSA['business_id'])]
dfrsSA_filtered

business_id                     city_state  \
9781   0x54d2ecba1ffb5c61:0x1af203277db3010b            Redding, California   
12288  0x80e852c733d0906d:0xc197a37de8c055e7            Ventura, California   
12364  0x872b75df478c253b:0x9172e57cd7e6887d            Scottsdale, Arizona   
15151  0x80dc0be615dfe3d1:0x50a4c215e36799e5           Carlsbad, California   
15495  0x876ea10d67911ff1:0xeb27e6ef5a647758              Greeley, Colorado   
15932  0x88f5c19c6dec7839:0xda0971b44367b12f                Dacula, Georgia   
16475  0x881992d75dd5a5d1:0x4dd25c13755f1d17              Zeeland, Michigan   
16901  0x4cb26bd192cce92d:0x83892a215d777b51                  Auburn, Maine   
17289  0x880f0ad897f4bd8d:0x8db335e4137a3e8c                 Cary, Illinois   
17790  0x87d8d1745cc3b8b3:0xd9929c0b8ff42c7e               Fenton, Missouri   
17851   0x89e4728792f8ac4b:0x511793a78ec5c2c         Milford, Massachusetts   
19407  0x87c8b904a7e90219:0x7fc0eb21d0f989ac               Joplin, Missouri   
19690  0x86d7fcbfb0650313:0x90a5a047e6d5d6e5              Thatcher, Arizona   
19913  0x876c91fa59dc9edf:0x9056441b9106059b               Parker, Colorado   
20159  0x89b7cd7615485661:0x8827692c66dfbbe5            Rockville, Maryland   
20510  0x88ee0d7df761e47f:0x4d6c32fd382bfe17                  Adel, Georgia   
21483  0x88f6728e170c34f3:0xa8395c203d9781a5                Athens, Georgia   
22124  0x89b7b9af2e9109a3:0x87443f600f805ad0  Washington, District Columbia   
22909  0x89e7c25e99fcd787:0x6b21ee0b5db8abbb         Naugatuck, Connecticut   
24063  0x88f590ea77a58539:0x8b6dc9dccea95793                Buford, Georgia   
24724  0x88f59e4b53b186a1:0x8b81d67b3112c3d0               Suwanee, Georgia   
24749  0x88f54f940b27e58f:0x16df3a2622163315          Cartersville, Georgia   
25080  0x872b754aa2f20781:0x49d1800ce30a9b37            Scottsdale, Arizona   
26278  0x89e36234ac5191df:0xf1c047d78cd85f4b        Weymouth, Massachusetts   
26294   0x89c81276e808b931:0x7095ded8be8f208         Cockeysville, Maryland   
26531  0x89e3868c93fb5a4d:0xe82743d4ca50d950          Natick, Massachusetts   
26926  0x89b075fbad2c7191:0x86bd736813fbf93d              Suffolk, Virginia   
27636  0x8803b8ba035e100f:0x4b9ac94cb9526b74              Neenah, Wisconsin   
27970  0x885683562ac7384d:0x44fbb2871fb84b08         Waxhaw, North Carolina   
27984  0x8840896b1c716dab:0x5a8de67b53b905a4               Miamisburg, Ohio   
28800  0x89c623f5095b2333:0x2b31b361ebb7253d        Lancaster, Pennsylvania   
29111  0x89c46a0b0af1bf11:0xee71ea7b7119119d             Bath, Pennsylvania   
29178  0x875286ac6e3e251b:0xeb8bd0721c884dc3                 Riverton, Utah   
29909  0x89b6ef0e74531869:0xe141f0a5ebafe4bb             Stafford, Virginia   
30377  0x884137a9f30e58f7:0x5b9fc78693c230fd                  Batavia, Ohio   
31027  0x86fc959f8989d62b:0xd65098c2a52b0160              Hobbs, New Mexico   
32230  0x8807a929e6188e53:0x61f65f5b8a9f9974           Middleton, Wisconsin   
32670  0x88536b9dbe0ea64d:0x1a1de1276cdb40ac      Randleman, North Carolina   
32804  0x89c6a506c47bc263:0xcac5c078c2398e52       Warminster, Pennsylvania   
33776  0x86346a0611eb9f4b:0x8c859a201ff1a0cc               Texarkana, Texas   
35750  0x89acf3107688a1c5:0xfb37f8600db6a79b           Cary, North Carolina   
35796  0x54950d2121223975:0x13e91c08bdd3f3ac                 Tigard, Oregon   
36423    0x87f7e1c200d1a317:0xa7bc9aff407831              Hudson, Wisconsin   
37105  0x89d3721334ec19d9:0x2cc1cf23a03db0f7            Tonawanda, New York   
37832  0x89c606065b13ce27:0x21d5bbf00aa2a29c        Myerstown, Pennsylvania   
38014  0x89ad1f02ff7da467:0x3ffcb90773a6346b         Durham, North Carolina   
38102  0x874da4b34598283b:0xe9602127bc8ef75a                    Salem, Utah   
38397  0x89ace5c724745963:0x47fceb8fd61fb090         Durham, North Carolina   
39768  0x89c85a3f241c95bb:0xa4bc77caa98ff02c          Hanover, Pennsylvania   

                                                  combSA  
9781   'Hot tub repair 

Sacamos las filas que contienen más palabras en la columna 'combSA'

In [234]:
# Dividir cada valor de 'combSA' en palabras y contar la cantidad de palabras
dfrsSA_filtered['word_count'] = dfrsSA_filtered['combSA'].str.split().apply(len)

# Encontrar la fila con el mayor número de palabras
fila_con_mas_palabras = dfrsSA_filtered[dfrsSA_filtered['word_count'] == dfrsSA_filtered['word_count'].max()]

# Mostrar la fila con más palabras en la columna 'combSA'
fila_con_mas_palabras

/var/folders/h3/j6yq8mwd3rl8gkw4cs7kj4p80000gn/T/ipykernel_80135/3108414760.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfrsSA_filtered['word_count'] = dfrsSA_filtered['combSA'].str.split().apply(len)


business_id                 city_state  \
22909  0x89e7c25e99fcd787:0x6b21ee0b5db8abbb     Naugatuck, Connecticut   
32670  0x88536b9dbe0ea64d:0x1a1de1276cdb40ac  Randleman, North Carolina   

                                                  combSA  word_count  
22909  'Appliance repair service' 'Identifies as wome...          45  
32670  'Appliance repair service' 'Identifies as wome...          45

Obtenemos el 'business_id' correspondiente a esa fila. Este 'business_id' será el input de nuestro sistema de recomendación.

In [235]:
fila_con_mas_palabras.iloc[0,0]

'0x89e7c25e99fcd787:0x6b21ee0b5db8abbb'

**Locality-Sensitive Hashing (LSH)** es una técnica utilizada en la minería de datos y la recuperación de información para aproximar la similitud entre puntos de datos en un espacio de alta dimensionalidad. A menudo se utiliza como una alternativa a las medidas tradicionales de similitud, como la similitud del coseno, cuando se trabaja con conjuntos de datos grandes en los que el cálculo de la similitud exacta se vuelve computacionalmente costoso.

Para este método, se utiliza algo llamado **Random Binary Projections**.

Las **Random Binary Projections** son un tipo específico de función hash utilizada en LSH. Estas funciones hash funcionan generando conjuntos de valores binarios (0 y 1) de manera aleatoria y utilizando estos valores para proyectar los puntos de datos en un espacio de menor dimensionalidad. La aleatoriedad de estas proyecciones es un factor clave en LSH, ya que introduce un nivel controlado de aleatoriedad que ayuda a agrupar puntos de datos similares con una probabilidad más alta.

Primero debemos descargar las librerias:

In [236]:
import numpy as np
from nearpy import Engine
from nearpy.hashes import RandomBinaryProjections
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

Al igual que en el modelo de similitud del coseno, necesitamos calcular los vectores y llenar los valores NaN con ''.

In [237]:
tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(dfrsSA_filtered['combSA'].fillna(''))

In [238]:
# Establecemos el número de dimensiones en la matriz TF-IDF
num_dimensions = tfidf_matrix.shape[1]

# También, establecemos el número de funciones de hash (proyecciones binarias aleatorias)
num_hash_functions = 4  # Podemos ajustar este valor según nuestras necesidades

# Creamos el motor LSH (Localización Sensible al Hashing)
engine = Engine(num_dimensions, lshashes=[RandomBinaryProjections('rbp', num_hash_functions)])

En el código anterior, podemos ver que hay un número de funciones de hash. Necesitamos calcular este hiperparámetro teniendo en cuenta este equilibrio:

**Mayor número de `num_hash_functions` (Más funciones de hash):**

- **Ventajas:**
    - **Mayor Precisión:** Con más funciones de hash, es menos probable que puntos de datos disímiles se asignen al mismo cubo. Esto puede llevar a una mayor precisión porque hay menos falsos positivos al buscar elementos similares.
- **Desventajas:**
    - **Menor Recuperación:** Mientras que la precisión puede aumentar, la recuperación (la capacidad de encontrar todos los elementos similares) puede disminuir. Más funciones de hash significan que los elementos similares deben coincidir en múltiples funciones de hash para considerarse similares, lo que podría resultar en que se omitan algunos elementos similares.

**Menor número de `num_hash_functions` (Menos funciones de hash):**

- **Ventajas:**
    - **Mayor Recuperación:** Con menos funciones de hash, es más fácil que puntos de datos similares se asignen al mismo cubo, aumentando las posibilidades de encontrar todos los elementos similares.
- **Desventajas:**
    - **Menor Precisión:** Sin embargo, un menor número de funciones de hash también puede resultar en más falsos positivos porque elementos no similares pueden asignarse accidentalmente al mismo cubo.


En este caso, hemos elegido 4. Veamos si tenemos una buena aproximación de lo que obtuvimos con el modelo de similitud del coseno.

Ahora, iteramos a través de las filas de la matriz TF-IDF y las asociamos con identificadores de elementos (IDs de elementos):

In [239]:
# Recorrer cada fila y su índice en la matriz TF-IDF
for i, row in enumerate(tfidf_matrix):

    # Extraer el 'business_id' de la fila correspondiente en el DataFrame 'df_FML1_for_Model'
    business_id = dfrsSA_filtered.iloc[i]['business_id']

    # Almacenar el vector TF-IDF como un arreglo aplanado en el motor LSH, asociado con el ID del negocio
    engine.store_vector(row.toarray().flatten(), data=business_id)

Hacemos la función que recomienda los 5 business_id que más se parecen al business_id elegido.

In [240]:
def Business_id_Recommendation(business_id, engine=engine):
    """
    Recommends similar games using Locality-Sensitive Hashing (LSH).

    Parameters:
        item_id (int): The ID of the game for which recommendations are requested.
        engine (LSH Engine): The LSH engine used for similarity search.

    Returns:
        list: A list of recommended games in JSON format.
    """
    
    # By entering the `item_id`, we should receive a list with 5 recommended games similar to the one entered.
    # We review the user enter a number
    if (type(business_id) == str):
        # Now, let's review the item_id exists
        if business_id in dfrsSA_filtered['business_id'].values:
                # We Get the LSH index of the input game

                # Query the TF-IDF matrix to get the TF-IDF vector for the input game
                query = tfidf_matrix[dfrsSA_filtered['business_id'] == business_id].toarray().flatten()

                # Use LSH to find similar games (neighbors) to the input game
                neighbors = engine.neighbours(query)

                # Recommendations based on LSH

                # Extract the game IDs of the recommended games, excluding the input game, and limit to the top 5
                recommended_business_ids = [neighbor[1] for neighbor in neighbors if neighbor[1] != business_id][:5]

                # Filter the DataFrame to get details of the recommended games (titles)
                recommended_business = dfrsSA_filtered[dfrsSA_filtered['business_id'].isin(recommended_business_ids)][['business_id']]

                # Construct the list of recommendations in JSON format
                result = [{'Rec {}'.format(i + 1): business} for i, business in enumerate(recommended_business['business_id'])]

                return result
        else:
            return print('The business_id does not exists. Please try again.')
    else:
        return print("The business_id must be a number, please enter a new item_id")

In [241]:
Business_id_Recommendation('0x89e7c25e99fcd787:0x6b21ee0b5db8abbb')

[{'Rec 1': '0x4cb26bd192cce92d:0x83892a215d777b51'},
 {'Rec 2': '0x89e4728792f8ac4b:0x511793a78ec5c2c'},
 {'Rec 3': '0x8803b8ba035e100f:0x4b9ac94cb9526b74'},
 {'Rec 4': '0x8840896b1c716dab:0x5a8de67b53b905a4'},
 {'Rec 5': '0x86fc959f8989d62b:0xd65098c2a52b0160'}]

In [242]:
def Business_id_Recommendation_combSA(business_id, engine=engine):
    """
    Recommends similar games using Locality-Sensitive Hashing (LSH).

    Parameters:
        item_id (int): The ID of the game for which recommendations are requested.
        engine (LSH Engine): The LSH engine used for similarity search.

    Returns:
        list: A list of recommended games in JSON format.
    """
    
    # By entering the `item_id`, we should receive a list with 5 recommended games similar to the one entered.
    # We review the user enter a number
    if (type(business_id) == str):
        # Now, let's review the item_id exists
        if business_id in dfrsSA_filtered['business_id'].values:
                # We Get the LSH index of the input game

                # Query the TF-IDF matrix to get the TF-IDF vector for the input game
                query = tfidf_matrix[dfrsSA_filtered['business_id'] == business_id].toarray().flatten()

                # Use LSH to find similar games (neighbors) to the input game
                neighbors = engine.neighbours(query)

                # Recommendations based on LSH

                # Extract the game IDs of the recommended games, excluding the input game, and limit to the top 5
                recommended_business_ids = [neighbor[1] for neighbor in neighbors if neighbor[1] != business_id][:5]

                # Filter the DataFrame to get details of the recommended games (titles)
                recommended_business = dfrsSA_filtered[dfrsSA_filtered['business_id'].isin(recommended_business_ids)][['combSA']]

                # Construct the list of recommendations in JSON format
                result = [{'Rec {}'.format(i + 1): business} for i, business in enumerate(recommended_business['combSA'])]

                return result
        else:
            return print('The business_id does not exists. Please try again.')
    else:
        return print("The business_id must be a number, please enter a new item_id")

In [243]:
Business_id_Recommendation_combSA('0x89e7c25e99fcd787:0x6b21ee0b5db8abbb')

[{'Rec 1': "'Appliance repair service' 'Curbside pickup' 'Delivery' 'In-store pickup' 'In-store shopping' 'Mask required' 'Staff required to disinfect surfaces between visits' 'Wheelchair accessible entrance' 'Assembly service' 'Repair services' 'Service guarantee' 'Debit cards'"},
 {'Rec 2': "'Appliance repair service' 'Curbside pickup' 'Delivery' 'In-store pickup' 'In-store shopping' 'Same-day delivery' 'LGBTQ friendly' 'Wheelchair accessible entrance' 'Wheelchair accessible parking lot' 'Assembly service' 'Repair services' 'Service guarantee' 'Quick visit' 'Checks' 'Debit cards' 'NFC mobile payments' 'Credit cards'"},
 {'Rec 3': "'Leather repair service' 'Identifies as women-led' 'Curbside pickup' 'In-store pickup' 'In-store shopping' 'Staff required to disinfect surfaces between visits' 'Wheelchair accessible entrance' 'Wheelchair accessible parking lot' 'Wheelchair accessible restroom' 'Quick visit' 'Checks' 'Debit cards' 'Credit cards'"},
 {'Rec 4': "'Appliance repair service ' '

In [244]:
data = [{'Rec 1': "Hot tub repair service 'Curbside pickup' 'Delivery' 'In-store pickup' 'In-store shopping' 'Mask required' 'Staff wear masks' 'Checks' 'Credit cards'"},
        {'Rec 2': "Leather repair service 'Identifies as women-led' 'Curbside pickup' 'In-store pickup' 'In-store shopping' 'Staff required to disinfect surfaces between visits' 'Wheelchair accessible entrance' 'Wheelchair accessible parking lot' 'Wheelchair accessible restroom' 'Quick visit' 'Checks' 'Debit cards' 'Credit cards'"},
        {'Rec 3': "Appliance repair service Refrigerator repair service 'Curbside pickup' 'Delivery' 'In-store pickup' 'In-store shopping' 'Mask required' 'Staff wear masks' 'Wheelchair accessible entrance' 'Assembly service' 'Service guarantee' 'Quick visit' 'Debit cards' 'Credit cards'"},
        {'Rec 4': "Leather repair service 'Mask required' 'Staff wear masks' 'Wheelchair accessible entrance' 'Appointments recommended'"},
        {'Rec 5': "Appliance repair service 'Identifies as women-led' 'Curbside pickup' 'Delivery' 'In-store pickup' 'In-store shopping' 'Same-day delivery' 'Mask required' 'Staff wear masks' 'Staff required to disinfect surfaces between visits' 'Wheelchair accessible entrance' 'Wheelchair accessible parking lot' 'Repair services' 'Service guarantee' 'Quick visit' 'Checks' 'Debit cards' 'Credit cards'"}]

# Inicializa una lista para almacenar las características únicas
caracteristicas = []

# Itera a través de los diccionarios y extrae las palabras
for item in data:
    for value in item.values():
        palabras = value.split("'")
        for palabra in palabras:
            palabra = palabra.strip()
            if palabra and palabra != ' ':
                caracteristicas.append(palabra)

# Convierte la lista de características en un conjunto para eliminar duplicados
caracteristicas_set = set(caracteristicas)

# Convierte el conjunto nuevamente en una lista
caracteristicas = list(caracteristicas_set)

# Ordena la lista alfabéticamente si es necesario
caracteristicas.sort()

# Crea un diccionario con la clave 'Características Recomendadas'
caracteristicas_dict = {'Características Recomendadas': caracteristicas}

# Muestra la lista de características únicas
print(caracteristicas_dict)


{'Características Recomendadas': ['Appliance repair service', 'Appliance repair service Refrigerator repair service', 'Appointments recommended', 'Assembly service', 'Checks', 'Credit cards', 'Curbside pickup', 'Debit cards', 'Delivery', 'Hot tub repair service', 'Identifies as women-led', 'In-store pickup', 'In-store shopping', 'Leather repair service', 'Mask required', 'Quick visit', 'Repair services', 'Same-day delivery', 'Service guarantee', 'Staff required to disinfect surfaces between visits', 'Staff wear masks', 'Wheelchair accessible entrance', 'Wheelchair accessible parking lot', 'Wheelchair accessible restroom']}


Ahora que ya vemos que el sistema de recomendación funciona, vamos a desarrollar una función mucho más general en el que el usuario elija un número del 1 al 13 para escoger la categoría de negocio y el sistema de recomendación le devuelva los atributos y servicios más recomendados.  

In [469]:
def Sistema_Recomendacion(num_categoria):

    """
    Basado en un área del negocio de contratistas en estados unidos ingresada por el usuario en forma de número entero,
    esta función recomienda diferentes caracteristicas que tienen las mejores empresas en esta área y las mejores ciudades
    para futura inversión.

    Las categorías son las siguientes:
    1. Appliance Repair & Installation.
    2. Cleaning Services.
    3. Interior Remodeling.
    4. HVAC & Air Quality. 
    5. Handyman or Other.
    6. Landscape & Garden Services. 
    7. Exterior Remodeling or Construction.
    8. Roofing & Insulation.
    9. Plumbing Services.
    10. Masonry & Concrete Services.
    11. Electrical Services.
    12. Home Inspection & Real Estate
    13. Art & Enterteinment

    Parámetros:
        num_categoria (int): Número de categoría que representa el área del negocio de contratistas.

    Retorna:
        dict: Un diccionario con las carácterísticas y ciudades enumeradas.
              Ejemplo: [{'Característica 1': 'Air conditioning contractor'},
                        {'Característica 2': 'Air conditioning repair service'},
                        {'Característica 3': 'Air duct cleaning service'},
                        {'Característica 4': 'Appointment required'},
                        {'Característica 5': 'HVAC contractor'},
                        {'Característica 6': 'Furnace repair service'},
                        {'Característica 7': 'Heating contractor'},
                        {'Característica 8': 'Mask required'},
                        {'Característica 9': 'Online estimates'},
                        {'Característica 10': 'Onsite services'},
                        {'Característica 11': 'Repair services'},
                        {'Característica 12': 'Staff get temperature checks'},
                        {'Característica 13': 'Staff required to disinfect surfaces between visits'},
                        {'Característica 14': 'Staff wear masks'},
                        {'Característica 15': 'Temperature check required'},
                        {'Ciudad 1': 'Brentwood, California'},
                        {'Ciudad 2': 'Gilbert, Arizona'},
                        {'Ciudad 3': 'Lisle, Illinois'},
                        {'Ciudad 4': 'Sarasota, Florida'}]
    """

    #Verificamos que el número de categoría ingresado es un número entero entre 1 y 11. No hay mejores negocios para las categorias 12
    # y 13, por lo que no se contemplan en esta función.
    if ((isinstance(num_categoria, int)) & ((num_categoria > 0) & (num_categoria < 14))):
        if ((num_categoria > 11) & (num_categoria < 14)):
            return print('Para esta categoría no hay buenos negocios que permitan ofrecer una recomendación')
        else:
            #Definimos el nombre de la categoría para filtrar en el dataframe llamado 'dfcrit' y 
            categoria = unique_categories[num_categoria-1]

            # 1. Vamos a recomendar los servicios y atributos que más se sugieren tener basados en la selección de la categoria de constratistas ingresada por el usuario.
            mask1 = (dfcrit['high_star_business'] == True) & (dfcrit['high_review_count'] == True) & (dfcrit['categories'] == categoria)
            bid_rsSA = dfcrit[mask1]
            
            # Ahora aplicamos este filtro a los mejores negocios.
            dfrsSA_filtered = dfrsSA[dfrsSA['business_id'].isin(bid_rsSA['business_id'])].copy()

            # Obtenemos la fila que contiene mayor número de palabras en la columna combSA, lo cual indica que es el business_id 
            # que ofrece más variedad de servicios y atributos. Esto permite que el sistema de recomendación sugiera negocios 
            # con mayor cantidad de servicios y atributos para que Construction Valdez los implemente. 
            # Dividir cada valor de 'combSA' en palabras y contar la cantidad de palabras
            dfrsSA_filtered['word_count'] = dfrsSA_filtered['combSA'].str.split().apply(len)

            # Encontrar la fila con el mayor número de palabras
            fila_con_mas_palabras = dfrsSA_filtered[dfrsSA_filtered['word_count'] == dfrsSA_filtered['word_count'].max()]

            # Obtenemos el business_id con más servicios y atributos para esa categoría.
            business_id = fila_con_mas_palabras.iloc[0,0]
            
            # Ahora vectorizamos las palabras de la columna dfrsSA_filtered['combSA']
            tfidf_vectorizer = TfidfVectorizer()
            tfidf_matrix = tfidf_vectorizer.fit_transform(dfrsSA_filtered['combSA'].fillna(''))

            # Establecemos el número de dimensiones en la matriz TF-IDF
            num_dimensions = tfidf_matrix.shape[1]

            # También, establecemos el número de funciones de hash (proyecciones binarias aleatorias)
            num_hash_functions = 4  # Podemos ajustar este valor según nuestras necesidades

            # Creamos el motor LSH (Localización Sensible al Hashing)
            engine = Engine(num_dimensions, lshashes=[RandomBinaryProjections('rbp', num_hash_functions)])
            
            # Recorrer cada fila y su índice en la matriz TF-IDF
            for i, row in enumerate(tfidf_matrix):

                # Extraer el 'business_id' de la fila correspondiente en el DataFrame 'dfrsSA_filtered'
                business_id1 = dfrsSA_filtered.iloc[i]['business_id']

                # Almacenar el vector TF-IDF como un arreglo aplanado en el motor LSH, asociado con el ID del negocio
                engine.store_vector(row.toarray().flatten(), data=business_id1)

            # Obtenemos el índice LSH del juego de entrada

            # Consultamos la matriz TF-IDF para obtener el vector TF-IDF del business_id de entrada
            query = tfidf_matrix[dfrsSA_filtered['business_id'] == business_id].toarray().flatten()

            # Usamos LSH para encontrar business_id similares (vecinos) al business_id de entrada
            neighbors = engine.neighbours(query)

            # Recomendaciones basadas en LSH

            # Extraemos los business_id de los business_id recomendados, excluyendo el business_id de entrada, y limitamos a los primeros 5
            recommended_business_ids = [neighbor[1] for neighbor in neighbors if neighbor[1] != business_id][:5]

            # Filtramos el DataFrame para obtener detalles de los business_id recomendados (combSA)
            recommended_business = dfrsSA_filtered[dfrsSA_filtered['business_id'].isin(recommended_business_ids)][['combSA']]

            # Construimos la lista de recomendaciones en formato JSON
            data = [{'Rec {}'.format(i + 1): business} for i, business in enumerate(recommended_business['combSA'])]

            # Inicializa una lista para almacenar las características únicas
            caracteristicas = []

            # Itera a través de los diccionarios y extrae las palabras
            for item in data:
                for value in item.values():
                    palabras = value.split("'")
                    for palabra in palabras:
                        palabra = palabra.strip()
                        if palabra and palabra != ' ':
                            caracteristicas.append(palabra)

            # Convierte la lista de características en un conjunto para eliminar duplicados
            caracteristicas_set = set(caracteristicas)

            # Convierte el conjunto nuevamente en una lista
            caracteristicas = list(caracteristicas_set)

            # Ordena la lista alfabéticamente si es necesario
            caracteristicas.sort()

            # Definir la cadena a buscar y la cadena de reemplazo
            cadena_buscar = 'C contractor'
            cadena_reemplazo = 'HVAC contractor'

            # Recorrer la lista y reemplazar 'C contractor' por 'HVAC contractor'
            for i in range(len(caracteristicas)):
                if caracteristicas[i] == cadena_buscar:
                    caracteristicas[i] = cadena_reemplazo

            # 2. Vamos a recomendar los servicios y atributos que más se sugieren tener basados en la selección de la categoria de constratistas ingresada por el usuario.
            mask1 = (dfcrit['high_category_consumption'] == True) & (dfcrit['low_category_volume'] == True) & (dfcrit['high_city_state_income'] == True) & (dfcrit['categories'] == categoria)
            bid_rsSA = dfcrit[mask1]

            # Ahora aplicamos este filtro a los mejores negocios.
            dfrsSA_filtered = dfrsSA[dfrsSA['business_id'].isin(bid_rsSA['business_id'])].copy()

            # Obtenemos la fila que contiene mayor número de palabras en la columna combSA, lo cual indica que es el business_id 
            # que ofrece más variedad de servicios y atributos. Esto permite que el sistema de recomendación sugiera negocios 
            # con mayor cantidad de servicios y atributos para que Construction Valdez los implemente. 
            # Dividir cada valor de 'combSA' en palabras y contar la cantidad de palabras
            dfrsSA_filtered['word_count'] = dfrsSA_filtered['combSA'].str.split().apply(len)

            # Encontrar la fila con el mayor número de palabras
            fila_con_mas_palabras = dfrsSA_filtered[dfrsSA_filtered['word_count'] == dfrsSA_filtered['word_count'].max()]

            # Obtenemos el business_id con más servicios y atributos para esa categoría.
            business_id = fila_con_mas_palabras.iloc[0,0]

            # Ahora vectorizamos las palabras de la columna dfrsSA_filtered['combSA']
            tfidf_vectorizer = TfidfVectorizer()
            tfidf_matrix = tfidf_vectorizer.fit_transform(dfrsSA_filtered['combSA'].fillna(''))

            # Establecemos el número de dimensiones en la matriz TF-IDF
            num_dimensions = tfidf_matrix.shape[1]

            # También, establecemos el número de funciones de hash (proyecciones binarias aleatorias)
            num_hash_functions = 4  # Podemos ajustar este valor según nuestras necesidades

            # Creamos el motor LSH (Localización Sensible al Hashing)
            engine = Engine(num_dimensions, lshashes=[RandomBinaryProjections('rbp', num_hash_functions)])

            # Recorrer cada fila y su índice en la matriz TF-IDF
            for i, row in enumerate(tfidf_matrix):

                # Extraer el 'business_id' de la fila correspondiente en el DataFrame 'dfrsSA_filtered'
                business_id1 = dfrsSA_filtered.iloc[i]['business_id']

                # Almacenar el vector TF-IDF como un arreglo aplanado en el motor LSH, asociado con el ID del negocio
                engine.store_vector(row.toarray().flatten(), data=business_id1)

            # Obtenemos el índice LSH del juego de entrada

            # Consultamos la matriz TF-IDF para obtener el vector TF-IDF del business_id de entrada
            query = tfidf_matrix[dfrsSA_filtered['business_id'] == business_id].toarray().flatten()

            # Usamos LSH para encontrar business_id similares (vecinos) al business_id de entrada
            neighbors = engine.neighbours(query)

            # Recomendaciones basadas en LSH

            # Extraemos los business_id de los business_id recomendados, excluyendo el business_id de entrada, y limitamos a los primeros 5
            recommended_business_ids = [neighbor[1] for neighbor in neighbors if neighbor[1] != business_id][:5]

            # Filtramos el DataFrame para obtener detalles de los business_id recomendados (city_state)
            recommended_business = dfrsSA_filtered[dfrsSA_filtered['business_id'].isin(recommended_business_ids)][['city_state']]

            # Construimos la lista de recomendaciones en formato JSON
            data = [{'Rec {}'.format(i + 1): business} for i, business in enumerate(recommended_business['city_state'])]

            # Inicializa una lista para almacenar las características únicas
            ciudades = []

            # Itera a través de los diccionarios y extrae las palabras
            for item in data:
                for value in item.values():
                    palabras = value.split("'")
                    for palabra in palabras:
                        palabra = palabra.strip()
                        if palabra and palabra != ' ':
                            ciudades.append(palabra)

            # Convierte la lista de características en un conjunto para eliminar duplicados
            ciudades_set = set(ciudades)

            # Convierte el conjunto nuevamente en una lista
            ciudades = list(ciudades_set)

            # Ordena la lista alfabéticamente si es necesario
            ciudades.sort()

            # Crea un diccionario con la clave 'Características Recomendadas'
            #dic_resultado = {'Características Recomendadas': caracteristicas, 'Ciudades Recomendadas': ciudades}

            #return print(dic_resultado)

            # Crear una lista de diccionarios para las características
            caracteristicas_formateadas = [{'Característica {}'.format(i + 1): caracteristica}
                                        for i, caracteristica in enumerate(caracteristicas)]

            # Crear una lista de diccionarios para las ciudades
            ciudades_formateadas = [{'Ciudad {}'.format(i + 1): ciudad}
                                    for i, ciudad in enumerate(ciudades)]

            # Combinar las listas de características y ciudades formateadas en una sola lista
            resultado_formateado = caracteristicas_formateadas + ciudades_formateadas

            return resultado_formateado
    else:
        #Mandamos un mensaje de error.
        return print('Por favor ingrese un número entero entre el 1 y el 13')


In [470]:
Sistema_Recomendacion(0)

Por favor ingrese un número entero entre el 1 y el 13
